druid-shell/src/platform/x11/application.rs

```diff
@@ -350,23 +369,62 @@ impl Application {
         Ok(false)
     }
 
-    pub fn run(self, _handler: Option<Box<dyn AppHandler>>) {
+    fn run_inner(self) -> Result<(), Error> {
+        // Try to figure out the refresh rate of the current screen. We run the idle loop at that
+        // rate.
+        let refresh_rate = util::refresh_rate(self.connection(), self.window_id).unwrap_or(60.0);
+        let timeout = Duration::from_millis((1000.0 / refresh_rate) as u64);
+        let mut last_idle_time = Instant::now();
```

Comment:
I'm not sure why we need this when using the present extension.
I suppose that's the reason why `poll_with_timeout` ignores its timeout when there is no idle action?

Reply:
The idea behind rate-limiting the idle loop is just to [avoid excessive wake-ups](https://wiki.ubuntu.com/Kernel/PowerManagement/IdentifyingIssues) by consolidating the idle commands and dong them all at once. It probably isn't that important, but I needed to support it anyway when present isn't being used. I've expanded on the comment a little.

Reply:
I see, but `poll_with_timeout` is already doing the waiting for external events, so why do we have to rate limit `run_idle` in this function separately? When there is an X11 event, why not just handle the idle queue as well?

Reply:
Hm, I guess that would make the most sense when using present, but when using the fallback it would mean redrawing on every event. Thinking more about it, it would make more sense for the fallback to use timers instead of the idle loop for drawing but I haven't got around to implementing timers yet... would it be ok if I left in a TODO comment?

I'm thinking that the ideal behavior for the idle loop would be to run it after every collection of events, or if there are no events then wait at least 1/refresh_rate before running the idle loop.

Reply:
> but when using the fallback it would mean redrawing on every event.
> Thinking more about it, it would make more sense for the fallback to use timers instead of the idle loop for drawing

Yeah, I think that would make more sense, delaying idle events just because we could be using a fallback render loop seems bad to me.

> would it be ok if I left in a TODO comment?

I would prefer an issue (or adding it to the existing timer issue), but yes.

> I'm thinking that the ideal behavior for the idle loop would be to run it after every collection of events, or if there are no events then wait at least 1/refresh_rate before running the idle loop.

You are referring to X11 events I suppose? In that case, I agree with this idea.